### DISCUSSION TITLE
issue: object converted to array when field is unregistered

### DISCUSSION THREAD

**Discussion by Ploppy3** [react-hook-form/react-hook-form#8316] [Q&A]
### Version Number

7.30.0

### Codesandbox/Expo snack

https://codesandbox.io/s/crimson-monad-tpicxn?file=/src/App.js

### Steps to reproduce

1.Click on 'clear'
2.The `names` field is now an array

### Expected behaviour

It should still be an object

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Ploppy3** [comment]
I am trying to remove a field from on object of my form.
I don't think I'm doing something right though.

What I want is to remove a field that is controlled by a component while keeping the component on screen:

```js
{
  features: {
    3: { /*value controlled*/ },
    5: { /*value controlled*/ },
  },
}
```

```js
{
  features: {
    5: { /*value controlled*/ },
  },
}
```
I tried to set the value to undefined but it does not work either, the value is removed from the field but `Controller` still provide the latest value.

I think that I should set the value to `null`. Can you please confirm that it is the right way to do it?

**Comment by bluebill1049** [comment]
> I think that I should set the value to null. Can you please confirm that it is the right way to do it?

That's correct.

why `undefined` is not supported.

- for the controlled component, it will result in the controlled becoming an uncontrolled warning message 
- we support inline defaultValue, undefined will allow us to check when to fall back to `defaultValues` at `useForm` level.

also, the number index is used for the array, this is by design so we can do a type check against your input name.

**Comment by Ploppy3** [comment]
> also, the number index is used for the array, this is by design so we can do a type check against your input name.

I'm not sure to understand, I thought the type would be determined based on `defaultValues`.

**Comment by bluebill1049** [comment]
> I'm not sure to understand, I thought the type would be determined based on defaultValues.

<img width="901" alt="Screen Shot 2022-05-06 at 11 11 45 pm" src="https://user-images.githubusercontent.com/10513364/167138118-f9e8aca1-c6ba-4958-815e-73e148c2458c.png">

not really, the index number should be reserved for the array index. This is by design so we have a consistent way to type check (TS) field path.

**Reply by bluebill1049** [reply]
unregister is basically performance a unset, it has no context on what's within the `defaultValues` as well

**Comment by Ploppy3** [comment]
So this is the expected behavior?

I just want to make sure to understand this.

If I want to enforce an object I have to make sure the keys are strings and not numbers? Otherwise a on object with keys as numbers could be recreated as an array as demonstrated above in the repro?

**Reply by bluebill1049** [reply]
that's pretty much the limitation with the current design by aligning strictly align with ts and js, which is the reason why we walk away from `test[0]` which would guarantee it's an array. It's not possible with my knowledge to get type-safe with `[]` syntax.

**Reply by bluebill1049** [reply]
- index number for array 
- string for object key
